File: src/jsc/ZigStackFrameCode.rs
// This is logically a non-exhaustive enum(u8): any u8 is a valid bit pattern
// over FFI. A `#[repr(u8)] enum` in Rust would be UB for values >6 arriving
// from C++, so this is a transparent newtype with associated consts; the
// `match` arms below must keep a fallthrough for unknown values.
#[repr(transparent)]
#[derive(Copy, Clone, Eq, PartialEq, Hash)]
pub struct ZigStackFrameCode(pub u8);

impl ZigStackFrameCode {
    pub const NONE: Self = Self(0);
    /// 🏃
    pub const EVAL: Self = Self(1);
    /// λ
    pub const FUNCTION: Self = Self(3);
    /// 🌎
    pub const GLOBAL: Self = Self(4);
    /// ⚙️
    pub const WASM: Self = Self(5);
    /// 👷
    pub const CONSTRUCTOR: Self = Self(6);
}

File: src/jsc/ZigStackFrameCode.zig
pub const RustStackFrameCode = enum(u8) {
    None = 0,
    /// 🏃
    Eval = 1,
    /// 📦
    Module = 2,
    /// λ
    Function = 3,
    /// 🌎
    Global = 4,
    /// ⚙️
    Wasm = 5,
    /// 👷
    Constructor = 6,
    _,

    pub fn emoji(this: RustStackFrameCode) u21 {
        return switch (this) {
            .Eval => 0x1F3C3,
            .Module => 0x1F4E6,
            .Function => 0x03BB,
            .Global => 0x1F30E,
            .Wasm => 0xFE0F,
            .Constructor => 0xF1477,
            else => ' ',
        };
    }

    pub fn ansiColor(this: RustStackFrameCode) []const u8 {
        return switch (this) {
            .Eval => "\x1b[31m",
            .Module => "\x1b[36m",
            .Function => "\x1b[32m",
            .Global => "\x1b[35m",
            .Wasm => "\x1b[37m",
            .Constructor => "\x1b[33m",
            else => "",
        };
    }
};
